encoding/multi_reader_iterator.go

```diff
@@ -0,0 +1,251 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package encoding
+
+import (
+	"container/heap"
+	"errors"
+	"io"
+
+	"github.com/m3db/m3db/interfaces/m3db"
+	xtime "github.com/m3db/m3db/x/time"
+)
+
+var (
+	errOutOfOrderIterator = errors.New("series values are out of order from mixed reader")
+)
+
+// multiReaderIterator is an iterator that iterates in order over a list of sets of
+// internally ordered but not collectively in order readers, it also deduplicates datapoints.
+type multiReaderIterator struct {
```

Comment:
Yea not so much about memory but more about performance, since using the heap adds a bit of overhead compared to just using the single reader, would be nice to clean it up but guess it probably doesn't matter that much if it's only one reader.